File: services/28-node-grpc/src/auth/interceptor.ts
// gRPC auth interceptor — validates JWT or API key from gRPC metadata.
//
// gRPC metadata is the equivalent of HTTP headers.
// Clients send: metadata.set('authorization', 'Bearer <token>')
//               metadata.set('x-api-key', '<key>')
//
// On valid JWT: attaches user to call context, proceeds.
// On valid API key: DB lookup, attaches user, proceeds.
// On missing or invalid credential: sends UNAUTHENTICATED status, halts call.

import {
  ServerInterceptingCall,
  ServerInterceptor,
  status,
} from '@grpc/grpc-js'
import jwt from 'jsonwebtoken'
import { prisma } from '../db/client.js'
import type { AuthUser } from '../middleware/auth.js'

const JWT_SECRET = process.env.JWT_SECRET ?? 'dev-secret-change-in-production'

export function verifyToken(token: string): AuthUser | null {
  try {
    const payload = jwt.verify(token, JWT_SECRET) as AuthUser
    return { id: payload.id, email: payload.email, name: payload.name }
  } catch {
    return null
  }
}

// authInterceptor is a gRPC server interceptor.
// Register at server creation: new Server({ interceptors: [authInterceptor] })
export const authInterceptor: ServerInterceptor = (methodDescriptor, call) => {
  return new ServerInterceptingCall(call, {
    start(metadata, listener, next) {
      const authValues = metadata.get('authorization')
      const apiKeyValues = metadata.get('x-api-key')

      // --- Attempt 1: Bearer JWT ---
      if (authValues.length > 0) {
        const authHeader = authValues[0] as string
        if (authHeader.startsWith('Bearer ')) {
          const token = authHeader.slice(7)
          const user = verifyToken(token)
          if (user) {
            // Attach user to metadata for downstream handlers.
            metadata.set('x-user-id', user.id)
            metadata.set('x-user-email', user.email)
            metadata.set('x-user-name', user.name)
            next(metadata, listener)
            return
          }
        }
        call.sendStatus({
          code: status.UNAUTHENTICATED,
          details: 'Invalid or expired JWT token',
        })
        return
      }

      // --- Attempt 2: X-API-Key header ---
      if (apiKeyValues.length > 0) {
        const apiKey = apiKeyValues[0] as string
        prisma.user
          .findUnique({ where: { apiKey } })
          .then((user) => {
            if (!user) {
              call.sendStatus({
                code: status.UNAUTHENTICATED,
                details: 'Invalid API key',
              })
              return
            }
            metadata.set('x-user-id', user.id)
            metadata.set('x-user-email', user.email)
            metadata.set('x-user-name', user.name)
            next(metadata, listener)
          })
          .catch(() => {
            call.sendStatus({
              code: status.INTERNAL,
              details: 'Auth check failed',
            })
          })
        return
      }

      // --- No credentials ---
      call.sendStatus({
        code: status.UNAUTHENTICATED,
        details: 'Authentication required. Provide Bearer token or x-api-key metadata.',
      })
    },
  })
}

File: services/28-python-grpc/src/auth/interceptor.py
"""gRPC auth interceptor — validates JWT or API key from gRPC metadata.

gRPC metadata is the equivalent of HTTP headers.
Clients send: metadata = [('authorization', 'Bearer <token>')]
              metadata = [('x-api-key', '<key>')]

On valid JWT: calls servicer_context.proceed(), user is available in context.
On valid API key: DB lookup, proceeds with user attached.
On missing or invalid credential: aborts with StatusCode.UNAUTHENTICATED.
"""

import grpc
from sqlalchemy.future import select
from sqlalchemy.orm import Session

from src.auth.jwt import verify_token
from src.db.models import User
from src.db.session import AsyncSessionLocal


class AuthInterceptor(grpc.aio.ServerInterceptor):
    """gRPC async server interceptor that enforces JWT or API key auth."""

    async def intercept_service_async(self, continuation, handler_call_details):
        metadata = dict(handler_call_details.invocation_metadata)

        # --- Attempt 1: Bearer JWT ---
        auth_header = metadata.get("authorization", "")
        if auth_header.startswith("Bearer "):
            token = auth_header[len("Bearer "):]
            payload = verify_token(token)
            if payload is not None:
                return await continuation(handler_call_details)
            return _unauthenticated_handler("Invalid or expired JWT token")

        # --- Attempt 2: X-API-Key ---
        api_key = metadata.get("x-api-key")
        if api_key:
            async with AsyncSessionLocal() as session:
                result = await session.execute(
                    select(User).where(User.api_key == api_key)
                )
                user = result.scalars().first()
            if user is not None:
                return await continuation(handler_call_details)
            return _unauthenticated_handler("Invalid API key")

        return _unauthenticated_handler(
            "Authentication required. Provide Bearer token or x-api-key metadata."
        )


def _unauthenticated_handler(message: str):
    """Returns a gRPC handler that immediately aborts with UNAUTHENTICATED."""

    async def handler(request, context):
        await context.abort(grpc.StatusCode.UNAUTHENTICATED, message)

    return grpc.unary_unary_rpc_method_handler(handler)
